active/0000-addon-instrumentation-experimental-hooks.md

```diff
@@ -0,0 +1,400 @@
+//    |- Babel_1
+//    |--|- Funnel
+//    |- Babel_2
+node.label.name === "TreeMerger";
+for (n of node.adjacentIterator()) {
+  console.log(n.label.name);
+}
+// prints
+//
+// Babel_1
+// Babel_2
+
+
+for (n of node.preOrderIterator(x => x.label.name === 'Babel_2')) {
+  console.log(n.label.name);
+}
+// prints
+//
+//  TreeMerger
+//  |- Babel_1
+```
+
+###### dfsIterator(until)
+
+Returns an iterator that yields every node in the subgraph sourced at this node.
+Nodes are yielded in depth-first order.  If the optional parameter `until` is
+passed, nodes for which `until` returns `true` will not be yielded, nor will
+nodes in their subgraph, unless those nodes are reachable by some other path.
+
+Example:
+```js
+// for a graph
+//  TreeMerger
+//    |- Babel_1
+//    |--|- Funnel
+//    |- Babel_2
+for (n of node.dfsIterator()) {
+  console.log(n.label.name);
+}
+// prints
+//
+// TreeMerger
+// Babel_1
+// Funnel
+// Babel_2
+```
+
+###### bfsIterator()
+
+Returns an iterator that yields every node in the subgraph sourced at this node.
+Nodes are yielded in breadth-first order.  If the optional parameter `until` is
+passed, nodes for which `until` returns `true` will not be yielded, nor will
+nodes in their subgraph, unless those nodes are reachable by some other path.
+
+
+Example:
+```js
+// for a tree
+//  TreeMerger
+//    |- Babel_1
+//    |--|- Funnel
+//    |- Babel_2
+for (n of node.bfsIterator()) {
+  console.log(n.label.name);
+}
+// prints
+//
+// TreeMerger
+// Babel_1
+// Babel_2
+// Funnel
+```
+
+###### statsIterator()
+
+Returns an iterator that yields `[name, value]` pairs of stat names and values.
+
+Example:
+```js
+  //  for a typical broccoli node 
+  for ([statName, statValue] of node.statsIterator()) {
+    console.log(statName, statValue);
+  }
+  // prints
+  //
+  // "time.self" 64232794
+  // "fs.statSync.count" 40
+  // "fs.statSync.time" 401232123
+  // ...
+  ```
+
+# How We Teach This
+
+This has no effect on day-to-day usage of ember-CLI.  It is a tool to help users
+monitor and analyze their build performance, so documentation and teaching
+belong primarily in `PERF_GUIDE.md`.  Having said that, we should also add a
+section to `https://ember-cli.com/extending/` and the API docs to make using
+this feature easier for addon authors and CLI power users.
+
+# Drawbacks
+
+* No drawbacks come to mind, besides the ever present issue of maintenance
+
+# Alternatives
+
+One alternative is the status quo: with `BROCCOLI_VIZ=1` users can output a file
+with a similar format that they can post-process offline.  Although this works
+for manual analysis, it is considerably more cumbersome for any automated system
+(such as ongoing monitoring of build performance).  It also does not include
+instrumentation outside of the build, most notably startup.
+
+# Unresolved questions
+
+- [heimdalljs-tree](https://github.com/heimdalljs/heimdalljs-tree) supports
+  `Symbol.Iterator`; should we commit to this as part of our API?
```

Comment:
Yeah this is why I'm a little torn.  On the one hand, when you want to visit the entire subgraph, it seems nice to do

```
for (n of graph) {
}
```

but I agree that in most cases being explicit about the iteration strategy is good